### PR TITLE
Adding GCS to GoogleJobStore

### PR DESCRIPTION
**Pull request by olsona**

### CODE REVIEW

File: extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleCloudExtensionModule.java [hunk 1]
```diff
@@ -110,6 +117,15 @@ Datastore getDatastore(@ProjectId String projectId, GoogleCredentials credential
         .getService();
   }
 
+  @Provides
+  @Singleton
+  Bucket getBucket(@ProjectId String projectId) {
+    Storage storage = StorageOptions.getDefaultInstance().getService();
+    // TODO: this assumes the bucket already exists, is that a fair assumption?
```
[the remark this answers is not in the document]
Comment (rtannenbaum) [reply extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleCloudExtensionModule.java]: Can get rid of TODO now, thanks

File: extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleJobStore.java [hunk 2]
```diff
@@ -327,10 +300,64 @@ private Key getDataKey(UUID jobId, String key) {
...
+  public void create(UUID jobId, String key, InputStream stream) {
+    String keyName = getDataKeyName(jobId, key);
+    // TODO: use result to determine success/failure
+    googleUserDataStore.create(keyName, stream);
+  }
+
+  @Override
+  public InputStream getStream(UUID jobId, String key) {
+    String keyName = getDataKeyName(jobId, key);
+    com.google.cloud.storage.Blob blob = googleUserDataStore.get(keyName);
```
Comment (rtannenbaum) [inline extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleJobStore.java]: there is still more blob logic here... can the UserDataStore just take in jobId and key and figure out the rest itself? including getDataKeyName, that seems like logic particular to UserDataStore

File: extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleUserDataStore.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+package org.datatransferproject.cloud.google;
+
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.Bucket;
+import com.google.inject.Inject;
+import java.io.InputStream;
+
+public class GoogleUserDataStore {
```
Comment (rtannenbaum) [inline extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleUserDataStore.java]: documentation

File: extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleJobStore.java [hunk 4]
```diff
@@ -327,10 +300,62 @@ private Key getDataKey(UUID jobId, String key) {
...
+
+  @Override
+  public void create(UUID jobId, String key, InputStream stream) {
+    // TODO: use result to determine success/failure
+    googleUserDataStore.create(jobId, key, stream);
+  }
+
+  @Override
+  public InputStream getStream(UUID jobId, String key) {
+    com.google.cloud.storage.Blob blob = googleUserDataStore.get(jobId, key);
```
Comment (rtannenbaum) [inline extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleJobStore.java]: this is still blob logic within GoogleJobStore, right? 

can all the blob logic be moved into UserDataStore?

File: extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleJobStore.java [hunk 5]
```diff
@@ -327,10 +300,62 @@ private Key getDataKey(UUID jobId, String key) {
     if (entity == null) {
       return null;
```
Comment (rtannenbaum) [inline extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleJobStore.java]: I thought we were moving all generic-type storage to the new UserDataStore (via pass thru in the existing JobStore interface), while PortabilityJob stuff (i.e. job metadata) would continue to be stored in DataStore (the dataStore variable). Is all the "type T" generic stuff == job temp data? +@holachuy

File: distributions/demo-google-deployment/bin/setup_gke_environment.sh [hunk 6]
```diff
@@ -309,6 +309,12 @@ echo "Created GCS bucket $GCS_BUCKET_NAME"
 print_step "Granting service account ${SERVICE_ACCOUNT} viewer privileges to 'app-data' bucket"
 gsutil acl -p ${PROJECT_ID} ch -u ${SERVICE_ACCOUNT}:R ${GCS_BUCKET_NAME}
 
+print_step "Creating GCS 'user-data' bucket for storing encrypted user data"
+BUCKET_NAME="user-data-$PROJECT_ID"
+GCS_BUCKET_NAME="gs://$BUCKET_NAME/"
+gsutil mb -p ${PROJECT_ID} ${GCS_BUCKET_NAME}
```
Comment (holachuy) [inline distributions/demo-google-deployment/bin/setup_gke_environment.sh]: What happens if it already exists? We should add a comment to clarify

File: extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleCloudExtensionModule.java [hunk 7]
```diff
@@ -110,6 +116,14 @@ Datastore getDatastore(@ProjectId String projectId, GoogleCredentials credential
         .getService();
   }
 
+  @Provides
+  @Singleton
+  Bucket getBucket(@ProjectId String projectId) {
+    Storage storage = StorageOptions.getDefaultInstance().getService();
+    String bucketId = "user-data-" + projectId;
```
Comment (holachuy) [inline extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleCloudExtensionModule.java]: Add comment:
// Must match BUCKET_NAME in setup_gke_environment.sh

File: extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleJobStore.java [hunk 8]
```diff
@@ -48,28 +49,81 @@
...
+ * A {@link JobStore} implementation based on Google Cloud Platform's Datastore.
+ */
 @Singleton
 public final class GoogleJobStore implements JobStore {
+
   private static final String KIND = "persistentKey";
   private static final String CREATED_FIELD = "created";
 
   private final Datastore datastore;
+  // Keeping googleTempFileStore in GoogleJobStore is temporary.  Eventually we plan to pull it
```
Comment (holachuy) [inline extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleJobStore.java]: How about TODO: Refactor tempFileStore into a separate interface

File: extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleJobStore.java [hunk 9]
```diff
@@ -48,28 +49,81 @@
...
+        builder.put(property, (Timestamp) entity.getTimestamp(property));
+      } else {
+        Blob blob = entity.getBlob(property);
+        Object obj = null;
+        try {
+          try (ObjectInputStream in = new ObjectInputStream(blob.asInputStream())) {
+            try {
+              obj = in.readObject();
+            } catch (ClassNotFoundException e) {
+              e.printStackTrace();
```
Comment (holachuy) [inline extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleJobStore.java]: Let's throw IO exception instead here and below.

File: extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleTempFileStore.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.channels.Channels;
+import java.util.UUID;
+import org.datatransferproject.types.transfer.models.DataModel;
+
+/**
+ * Class for temporarily storing user data for transfer
+ */
+public class GoogleTempFileStore {
```
Comment (holachuy) [inline extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleTempFileStore.java]: // TODO: extract a temp file store interface

File: extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleJobStore.java [hunk 11]
```diff
@@ -199,12 +199,17 @@ public void remove(UUID jobId) throws IOException {
...
+  public PortabilityJob findJob(UUID jobId) {
     Entity entity = datastore.get(getKey(jobId));
     if (entity == null) {
       return null;
     }
-    return PortabilityJob.fromMap(getProperties(entity));
+    try {
+      return PortabilityJob.fromMap(getProperties(entity));
+    } catch (IOException | ClassNotFoundException e) {
+      // TODO: is this the right thing to do?
```
Comment (holachuy) [inline extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleJobStore.java]: Probably more like TODO: Rethrow as IOException and propagate to callers

**Pull request merged**